extensions/zig_demo/README.md

```diff
@@ -0,0 +1,5 @@
+# Ziglang demo extension
+
+This is an example of an ABI-compatible module written in [zig](https://ziglang.org/).
```

Comment:
1. The purpose is to educate istio users how to write extensions. Istio users hang around istio community so it makes to place examples in istio org.

2. It's easier to get started without any dependencies. Copying a couple of hundred of lines is fine. Only 5 lines are needed, the rest could be deleted depending on usage. The example has more stubs than needed.

Reply:
1. Couldn't you educate users to use Proxy-Wasm extensions? Otherwise, they'll be stuck with the few examples here without knowing that they can use other extensions and SDKs.

2. Implementing ABI directly is not more developer-friendly than using SDK, which would bring the example down to ~20 lines instead and make it much more readable, e.g.:
```
const std = @import("proxy-wasm");

pub fn proxy_on_log() !void {
    proxy-wasm.log(.info, "log {}", .{id});
    proxy-wasm.log(.info, "request size {}, with headers {}, header value {}, which is same as {}", .{
        proxy-wasm.getProperty(i64, .{ "request", "size" }),
        proxy-wasm.getProperty(i64, .{ "request", "total_size" }),
        proxy-wasm.getHeader(.request_headers, "wasm-test"),
        proxy-wasm.getProperty([]const u8, .{ "request", "headers", "wasm-test" }),
    });
    var map_buffer: [4096]u8 = undefined;
    var allocator = &std.heap.FixedBufferAllocator.init(&map_buffer).allocator;
    var map = std.StringHashMap([]const u8).init(allocator);
    _ = proxy-wasm.getAllHeaders(.request_headers, &map);
    var it = map.iterator();
    while (it.next()) |next| {
        proxy-wasm.log(.info, "{}: {}", .{ next.key, next.value });
    }
}
```

Reply:
Everything is too immature to be a library, zig, proxy-wasm, wasm itself. It just adds a burden of maintenance, when a self-contained script is sufficient. The main savings is really decoding and unwrapping slices, which are pretty minor.

Reply:
Agreed, but if product maturity is a requirement, then we shouldn't be using Proxy-Wasm or Envoy in the first place.

In any case, I think that Zig SDK would be useful, so if you don't want to do it, do you mind if I use this code as a base for it?

Reply:
Go at it. Zig is unstable, so you might need to keep up with language changes.

It would be nice to have a nullVM version of the shim. It would permit running on x86, and it seems doable since C is first class. There're some alignment and usize modifications necessary.